editoast/simulation/src/lib.rs

```diff
@@ -1,3 +1,341 @@
+/// - [trait SimulationStreamExt] to batch simulations (requires additional [type Simulation::Context] bounds)
+pub trait Simulation: Sized + Send {
+    /// Simulation output
+    type Output: DeserializeOwned + Serialize + Send;
+    /// Computation error when performing the simulation for cache misses
+    type Error: std::error::Error + Send;
+    /// Context required for simulation
+    ///
+    /// Additional bounds are required on this to batch simulation requests, cf. [trait SimulationStreamExt].
+    type Context: Send;
+
+    /// Number of cache entry read attempts per batch
+    ///
+    /// Customize this value based on the size of simulation results considering all reads hit.
+    const CACHE_READS_BATCH_SIZE: usize = 50; // 🫳🎩
+
+    /// Computes the cache key based on simulation inputs
+    fn key(&self, app_version: &str) -> String;
+
+    /// Performs the simulation according to inputs and context
+    ///
+    /// This function does **not** need to handle any caching concerns.
+    fn compute(
+        self,
+        ctx: Self::Context,
+    ) -> impl Future<Output = Result<Self::Output, Self::Error>> + Send;
+
+    /// Retrieves the simulation result from cache or performs the simulation
+    ///
+    /// To batch simulations, check out [trait SimulationStreamExt].
+    ///
+    /// # Errors
+    ///
+    /// Should any caching error occur while reading, the simulation is performed.
+    /// Cache write errors are ignored. So are serde errors for caching.
+    /// All errors are logged.
+    fn simulate(
+        self,
+        mut vkconn: cache::Connection,
+        ctx: Self::Context,
+    ) -> impl Future<Output = Result<Self::Output, Self::Error>> {
+        async move {
+            let key = self.key(vkconn.app_version());
+            if let Some(value) = match vkconn.json_get::<Self::Output, _>(&key).await {
+                Ok(v) => v,
+                Err(e) => {
+                    tracing::error!(?e, key, "cache read error — performing simulation");
+                    None
+                }
+            } {
+                Ok(value)
+            } else {
+                let value = self.compute(ctx).await?;
+                match serde_json::to_string(&value) {
+                    Err(e) => {
+                        tracing::error!(
+                            ?e,
+                            key,
+                            "serialization error before cache write — skipping cache write"
+                        );
+                        #[cfg(test)]
+                        panic!("Simulation::simulate: serialization error {key} {e:?}");
+                    }
+                    Ok(serialized) => {
+                        tokio::spawn(async move {
+                            use deadpool_redis::redis::AsyncCommands as _;
+                            if let Err(e) = vkconn.set::<_, _, ()>(key, serialized).await {
+                                tracing::error!(?e, "cache write error");
```

Comment:
It might be useful to put the `key` in the logs, what do you think?

Reply:
Do you think we would have enough information from the span context to understand which cache write failed? And I'd say that cloning a key is probably negligible compare to a connection with valkey 🤷 But I'm fine with whatever the decision you take. I'll let you resolve.